File: matlab/examples_paper/algorithms/AAD.m
function [estimate, x, y] = AAD(S)
% Averages diagonals. Note that x(1) is NOT scaled by 0.5.
n = size(S, 1) - 1;
z = zeros(n+1, 1);
for k = 0:n
   z(k+1) = mean(diag(S, k)); 
end
x = real(z);
y = imag(z(2:n+1));
estimate = toeplitz([x(1); x(2:end)+1i*y]);
end

File: python/examples_paper/algorithms/aad.py
"""Averaging Along Diagonals (AAD) baseline estimator."""
import numpy as np
from scipy.linalg import toeplitz


def aad(S):
    """Estimate Toeplitz covariance by averaging diagonals of S.

    Parameters
    ----------
    S : ndarray, shape (n, n)
        Sample covariance matrix.

    Returns
    -------
    dict with keys 'estimate', 'x', 'y'.
    """
    n = S.shape[0]
    z = np.array([np.mean(np.diagonal(S, k)) for k in range(n)])
    x = np.real(z)
    y = np.imag(z[1:])
    estimate = toeplitz(z)
    return {"estimate": estimate, "x": x, "y": y}
